File: LC/323.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        parent = [i for i in range(n)]
        rank = [1] * (n)
        
        def find(i):
            p = parent[i]
            
            while p != parent[p]:
                parent[p] = parent[parent[p]]
                p = parent[p]
            
            return p
        
        def union(n1, n2):
            p1, p2 = find(n1), find(n2)
            
            if p1 == p2:
                return 0
            
            if rank[p1] > rank[p2]:
                parent[p2] = p1
                rank[p1] += rank[p2]
            else:
                parent[p1] = p2
                rank[p2] += rank[p1]
            
            return 1
        
        ans = n
        for e1, e2 in edges:
            ans -= union(e1, e2)
        
        return ans

File: LC/C++/Medium/323.cpp
class UnionFind {
public:
    UnionFind(int n) : parent_(n), rank_(n, 0) {
        for (size_t i{}; i < n; ++i) {
            parent_[i] = i;
        }
    }

    int find(int x) {
        while (x != parent_[x]) {
            parent_[x] = parent_[parent_[x]];
            x = parent_[x];
        }

        return x;
    }

    int unite(int a, int b) {
        a = find(a);
        b = find(b);

        if (a == b) return 0;

        parent_[b] = a;

        if (rank_[a] < rank_[b]) {
            std::swap(a, b);
        } else if (rank_[a] == rank_[b]) {
            rank_[a]++;
        }

        return 1;
    }

private:
    std::vector<int> parent_;
    std::vector<int> rank_;
};

class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        UnionFind uf(n);
        for (auto& edge : edges) {
            n -= uf.unite(edge[0], edge[1]);
        }

        return n;
    }
};